ASoC: SOF: (Intel HDA) Add support for DSPless debug mode

ujfalusi opened the pull request:
Hi,

this series will enable the use of the SOF Linux stack without DSP offloading.

In this mode no firmware loading will happen, the topology parsing is going to be reduced to only look for the DAI widget(s) and the IPC dependent callbacks are going to be ignored.

This mode can give another level of hardware verification on platforms where the DSP can be ignored and the audio interfaces can be tested directly.

On Intel platforms we can use this mode to verify programming flows against the legacy HDA stack if there is a need to debug in that level and don't have the DSP programming sequences interfering.

The use of DSPless mode is governed by the SOF_DBG_DSPLESS_MODE flag which is only going to be set if the user sets sof_debg=0x8000 and the platform advertises that the DSPless mode is supported (tested) on them.

ujfalusi commented:
Changes since v1:
- 32bit audio playback works correctly

ujfalusi commented:
@jsarha, fyi

plbossart commented:
@ujfalusi can we remove the WIP patches?

````
https://github.com/thesofproject/linux/commit/5457428f0758b05ee705c79ccaaafe29bd0a4694 acpi: HACK: silence the Found %d idle state
https://github.com/thesofproject/linux/commit/bbf0d7beb26eb74c661aaf1c0c542450a65e676c [WIP] snd_soc_acpi_mach link_mask to mach_params.link_mask stuff
https://github.com/thesofproject/linux/commit/8a0c34d80583748ca29818e42c1eb413ce058738 [HACK] ASoC: SOF: ipc4-topology: Load one library for testing
https://github.com/thesofproject/linux/commit/deba7bc78f5471f94a6cc63025057d83203e520f Revert "[HACK] ASoC: SOF: ipc4-topology: Load one library for testing"
https://github.com/thesofproject/linux/commit/45c51d2770a3a49061c565c0d84bd0da5980d1d1 --- Hacks ---
https://github.com/thesofproject/linux/commit/f8679ff00cc4387f6547cc340fcdf53ad677f680 ASoC: SOF: sof-priv: Remove unused SOF_DAI_STREAM() and SOF_FORMATS
https://github.com/thesofproject/linux/commit/cfeefcdb0d6508b2a293a39e8945446abaa44f85 -- misc fixes --
https://github.com/thesofproject/linux/commit/ab2e04d8c7296f05b5150dda0f3d8972cb19332f ALSA: intel-nhlt: Change struct nhlt_acpi_table.desc struct array to u8
https://github.com/thesofproject/linux/commit/26c3ad357c7d2fb4303d5f0e0992dbceaa3f7be6 ALSA: intel-nhlt: Change struct nhlt_fmt.fmt_config struct array to u8
https://github.com/thesofproject/linux/commit/9129c00cfe4f33c825c2511cd03a0952ed484e8c ASoC: SOF: uapi: header: Convert struct sof_manifest.items[] to u8 array
https://github.com/thesofproject/linux/commit/a380f80538e965a87bb95d8ebd0aebc7d4aa1f68 ASoC: SOF: core: Print out the value of sof_debug if it is set
https://github.com/thesofproject/linux/commit/49eefb2c39981192703234182995664960d34a80 -- flex to u8 --
````

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -283,7 +283,7 @@ sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widg
 
 	/* return if the widget is in use or if it is already unprepared */
-	if (!swidget->prepared || swidget->use_count > 1)
+	if (!swidget || !swidget->prepared || swidget->use_count > 1)
```
typo in commit title: prepare

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -208,6 +208,11 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	/* set up platform component driver */
 	snd_sof_new_platform_drv(sdev);
 
+	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE)) {
```
nit-pick: would the code not be more readable with a boolean flag, e.g. sdev->dspless_selected. I see many many repeats of this 30-odd string, it's a bit of an eyesore.

ujfalusi replied:
I did not wanted to introduce a new boolean in sdev, but the clearing of the bit in `sof_core_debug` might not be a nice thing?
Let me check if there is anything preventing this.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -677,7 +677,8 @@ void snd_sof_new_platform_drv(struct snd_sof_dev *sdev)
 	pd->pcm_construct = sof_pcm_new;
 	pd->ignore_machine = drv_name;
-	pd->be_hw_params_fixup = sof_pcm_dai_link_fixup;
+	if (!sof_debug_check_flag(SOF_DBG_DSPLESS_MODE))
```
maybe add a comment here to explain why we don't have a fixup?

ujfalusi replied:
OK

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2272,7 +2352,11 @@ int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
+		ret = snd_soc_tplg_component_load(scomp, &sof_dspless_tplg_ops, fw);
+	else
+		ret = snd_soc_tplg_component_load(scomp, &sof_tplg_ops, fw);
+
```
nit-pick: use a temp pointer and have a single call to tplg_component_load?

ujfalusi replied:
I think this is clear enough, I don's see the benefit in this particular case.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -603,6 +603,88 @@ static const struct snd_soc_dai_ops ipc4_hda_dai_ops = {
 	.prepare = hda_dai_prepare,
 };
 
+static int dspless_hda_dai_hw_params(struct snd_pcm_substream *substream,
```
commit message typo:
sof_debug=0x8000

not sure what "supported (tested)" means?
did you mean "the platform
advertises its ability to transfer audio without the DSP enabled" ?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -603,6 +603,88 @@ static const struct snd_soc_dai_ops ipc4_hda_dai_ops = {
+	if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_PLAYBACK) {
+		snd_hdac_ext_bus_link_set_stream_id(hlink, hstream->stream_tag);
+	} else {
+	}
```
remove these empty-brackets or add a comment?

ujfalusi replied:
Oh, remained from debugging.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -603,6 +603,88 @@ static const struct snd_soc_dai_ops ipc4_hda_dai_ops = {
+	list_for_each_entry(s, &bus->stream_list, list) {
+		hext_stream = stream_to_hdac_ext_stream(s);
+
+		if (hext_stream->link_substream)
```
do we need this condition? link_prepared is always set unconditionally

ujfalusi replied:
true

plbossart commented on sound/soc/sof/intel/pci-apl.c:
```diff
@@ -29,6 +29,7 @@ static const struct sof_dev_desc bxt_desc = {
 	.chip_info = &apl_chip_info,
 	.ipc_supported_mask	= BIT(SOF_IPC) | BIT(SOF_INTEL_IPC4),
 	.ipc_default		= SOF_IPC,
+	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
```
commit title:spurious initial space

ujfalusi replied:
OK, let me check it

plbossart reviewed (commented):
thanks @ujfalusi, set of cosmetic comments below

RanderWang commented on sound/soc/sof/intel/hda-stream.c:
```diff
@@ -480,7 +480,7 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 	int ret;
 	u32 dma_start = SOF_HDA_SD_CTL_DMA_START;
-	u32 mask;
+	u32 mask = 0x1 << hstream->index;
```
BIT(hstream->index) ?

ujfalusi replied:
it is a code move, but sure, let's do that.

RanderWang commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1046,6 +1046,9 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
 #endif
 
+	if (sof_debug_check_flag(SOF_DBG_DSPLESS_MODE))
+		hdev->no_ipc_position = 1;
```
without this if,  hdev->no_ipc_position would be 0 ?

ujfalusi replied:
Probably some shared code would try to send IPC? But in any case, since we don't use DSP the ipc_position is not even a possibility.

RanderWang commented on sound/soc/sof/intel/pci-tgl.c:
```diff
@@ -61,6 +62,7 @@ static const struct sof_dev_desc tglh_desc = {
 	.ipc_default		= SOF_IPC,
+	.dspless_mode_supported	= true,		/* Only supported for HDaudio */
 	.default_fw_path = {
 		[SOF_IPC] = "intel/sof",
```
in dspless mode,  which topology will be used ?

ujfalusi replied:
Depending on the ipc_type, works with both IPC3 or IPC4 and there should not be any IPC dependency

ujfalusi replied:
But yes, we are loading the topology file to get the DAIs for the PCMs.

ujfalusi commented:
> @ujfalusi can we remove the WIP patches?
> 
> ```
> https://github.com/thesofproject/linux/commit/5457428f0758b05ee705c79ccaaafe29bd0a4694 acpi: HACK: silence the Found %d idle state
> https://github.com/thesofproject/linux/commit/bbf0d7beb26eb74c661aaf1c0c542450a65e676c [WIP] snd_soc_acpi_mach link_mask to mach_params.link_mask stuff
> https://github.com/thesofproject/linux/commit/8a0c34d80583748ca29818e42c1eb413ce058738 [HACK] ASoC: SOF: ipc4-topology: Load one library for testing
> https://github.com/thesofproject/linux/commit/deba7bc78f5471f94a6cc63025057d83203e520f Revert "[HACK] ASoC: SOF: ipc4-topology: Load one library for testing"
> https://github.com/thesofproject/linux/commit/45c51d2770a3a49061c565c0d84bd0da5980d1d1 --- Hacks ---
> https://github.com/thesofproject/linux/commit/f8679ff00cc4387f6547cc340fcdf53ad677f680 ASoC: SOF: sof-priv: Remove unused SOF_DAI_STREAM() and SOF_FORMATS
> https://github.com/thesofproject/linux/commit/cfeefcdb0d6508b2a293a39e8945446abaa44f85 -- misc fixes --
> https://github.com/thesofproject/linux/commit/ab2e04d8c7296f05b5150dda0f3d8972cb19332f ALSA: intel-nhlt: Change struct nhlt_acpi_table.desc struct array to u8
> https://github.com/thesofproject/linux/commit/26c3ad357c7d2fb4303d5f0e0992dbceaa3f7be6 ALSA: intel-nhlt: Change struct nhlt_fmt.fmt_config struct array to u8
> https://github.com/thesofproject/linux/commit/9129c00cfe4f33c825c2511cd03a0952ed484e8c ASoC: SOF: uapi: header: Convert struct sof_manifest.items[] to u8 array
> https://github.com/thesofproject/linux/commit/a380f80538e965a87bb95d8ebd0aebc7d4aa1f68 ASoC: SOF: core: Print out the value of sof_debug if it is set
> https://github.com/thesofproject/linux/commit/49eefb2c39981192703234182995664960d34a80 -- flex to u8 --
> ```

I have not removed them??

ujfalusi commented:
> @ujfalusi can we remove the WIP patches?
> 
> ```
> https://github.com/thesofproject/linux/commit/5457428f0758b05ee705c79ccaaafe29bd0a4694 acpi: HACK: silence the Found %d idle state
> https://github.com/thesofproject/linux/commit/bbf0d7beb26eb74c661aaf1c0c542450a65e676c [WIP] snd_soc_acpi_mach link_mask to mach_params.link_mask stuff
> https://github.com/thesofproject/linux/commit/8a0c34d80583748ca29818e42c1eb413ce058738 [HACK] ASoC: SOF: ipc4-topology: Load one library for testing
> https://github.com/thesofproject/linux/commit/deba7bc78f5471f94a6cc63025057d83203e520f Revert "[HACK] ASoC: SOF: ipc4-topology: Load one library for testing"
> https://github.com/thesofproject/linux/commit/45c51d2770a3a49061c565c0d84bd0da5980d1d1 --- Hacks ---
> https://github.com/thesofproject/linux/commit/f8679ff00cc4387f6547cc340fcdf53ad677f680 ASoC: SOF: sof-priv: Remove unused SOF_DAI_STREAM() and SOF_FORMATS
> https://github.com/thesofproject/linux/commit/cfeefcdb0d6508b2a293a39e8945446abaa44f85 -- misc fixes --
> https://github.com/thesofproject/linux/commit/ab2e04d8c7296f05b5150dda0f3d8972cb19332f ALSA: intel-nhlt: Change struct nhlt_acpi_table.desc struct array to u8
> https://github.com/thesofproject/linux/commit/26c3ad357c7d2fb4303d5f0e0992dbceaa3f7be6 ALSA: intel-nhlt: Change struct nhlt_fmt.fmt_config struct array to u8
> https://github.com/thesofproject/linux/commit/9129c00cfe4f33c825c2511cd03a0952ed484e8c ASoC: SOF: uapi: header: Convert struct sof_manifest.items[] to u8 array
> https://github.com/thesofproject/linux/commit/a380f80538e965a87bb95d8ebd0aebc7d4aa1f68 ASoC: SOF: core: Print out the value of sof_debug if it is set
> https://github.com/thesofproject/linux/commit/49eefb2c39981192703234182995664960d34a80 -- flex to u8 --
> ```

Right, I did have a clean branch, but by mistake I opened the PR on my WIP branch, I will close this and open a new one, addressing the comments so far.
